# 高校数学とJulia言語 Day 2

- 城北中学校・高等学校　中学3年・高校1年
- 夏期講習会III 2026/8/24~2026/8/28
- 担当：清水団

## 本日のテーマ:整数問題をプログラムで考えよう!

### 今年のテーマ「実験する高校数学」

- 整数は**探す**・場合の数は**全部作る**・関数は**描く**・確率は**試す**

### 5日間の学習予定

- **Day 1**：Google Colabの紹介・Juliaで計算しよう ✅
- **Day 2**：整数問題をプログラムで考えよう ← 今日
- **Day 3**：場合の数・組合せを実験しよう
- **Day 4**：関数・グラフ・最大最小
- **Day 5**：確率・シミュレーション


## 今日のテーマ：「数学は実験できる」

整数の世界は、コンピュータでの**実験**にぴったりの分野です。

- 約数・素数・最大公約数 … 教科書の題材を「全部探す」
- 完全数・双子素数・フェルマー数 … 数学者が愛した数たちを「発見する」
- コラッツ予想 … **誰も解けていない未解決問題**を自分の手で「確かめる」

昨日のfor文とif文が今日の主役です!

## 約数を全部求めよう

$n$ の約数とは、$n$ を割り切る正の整数のこと。「$n \% d == 0$ となる $d$ を探す」だけです。

In [1]:
# 24の約数を全部表示
for d in 1:24
    if 24 % d == 0
        println(d)
    end
end

1
2
3
4
6
8
12
24


In [2]:
# 関数にしてしまおう! 約数のリストを返す
divisors(n) = [d for d in 1:n if n % d == 0]

divisors(24)

8-element Vector{Int64}:
  1
  2
  3
  4
  6
  8
 12
 24

In [3]:
# 約数の個数
length(divisors(24))

8

In [4]:
# 約数の和
sum(divisors(24))

60

In [5]:
# いろいろな数で実験してみよう
for n in [12, 36, 60, 100]
    println("$n の約数: $(divisors(n))")
end

12 の約数: [1, 2, 3, 4, 6, 12]
36 の約数: [1, 2, 3, 4, 6, 9, 12, 18, 36]
60 の約数: [1, 2, 3, 4, 5, 6, 10, 12, 15, 20, 30, 60]
100 の約数: [1, 2, 4, 5, 10, 20, 25, 50, 100]


## 素数判定

素数とは「1と自分自身以外に約数を持たない、2以上の整数」。つまり**約数がちょうど2個**の数です。

In [6]:
# 約数の個数が2個なら素数!
isprime_mine(n) = n ≥ 2 && length(divisors(n)) == 2

isprime_mine(17), isprime_mine(21)

(true, false)

In [7]:
# 1から30までの素数を表示
for n in 1:30
    if isprime_mine(n)
        println(n)
    end
end

2
3
5
7
11
13
17
19
23
29


## Primes.jl パッケージ

Juliaには素数専用のパッケージがあります。プロの道具を使ってみましょう。

In [8]:
# パッケージの追加（初回だけ少し時間がかかります）
using Pkg
Pkg.add("Primes")
using Primes

println("パッケージの読み込み完了!")

    Updating registry at `~/.julia/registries/General.toml`
   Resolving package versions...
     Project No packages added to or removed from `~/.julia/environments/v1.12/Project.toml`
    Manifest No packages added to or removed from `~/.julia/environments/v1.12/Manifest.toml`


パッケージの読み込み完了!


In [9]:
# 素数判定
isprime(9999999967)

true

In [10]:
# 100以下の素数を全部
primes(100)

25-element Vector{Int64}:
  2
  3
  5
  7
 11
 13
 17
 19
 23
 29
  ⋮
 59
 61
 67
 71
 73
 79
 83
 89
 97

In [11]:
# 100以下の素数は何個?
length(primes(100))

25

In [12]:
# 素因数分解も一発!
factor(2026)

2 * 1013

## エラトステネスの篩（ふるい）

紀元前3世紀の数学者エラトステネスが考えた、素数を**ふるい落として**見つける方法です。

1. 2から $n$ までの数を並べる
2. 残っている最小の数（素数）を残し、その倍数を全部消す
3. 繰り返す

In [13]:
# エラトステネスの篩を実装してみよう
function eratosthenes(n)
    is_prime = trues(n)      # 全部「素数候補」でスタート
    is_prime[1] = false      # 1は素数ではない
    for p in 2:n
        if is_prime[p]
            for q in 2p:p:n  # pの倍数をふるい落とす
                is_prime[q] = false
            end
        end
    end
    return [p for p in 1:n if is_prime[p]]
end

eratosthenes(100)

25-element Vector{Int64}:
  2
  3
  5
  7
 11
 13
 17
 19
 23
 29
  ⋮
 59
 61
 67
 71
 73
 79
 83
 89
 97

In [14]:
# Primes.jlの結果と一致するか確認
eratosthenes(1000) == primes(1000)

true

In [15]:
# 素数はどのくらいの割合であるのか実験
for n in [100, 1000, 10000, 100000]
    println("$n 以下の素数: $(length(primes(n))) 個（割合 $(round(length(primes(n))/n*100, digits=2))%）")
end

100 以下の素数: 25 個（割合 25.0%）
1000 以下の素数: 168 個（割合 16.8%）
10000 以下の素数: 1229 個（割合 12.29%）
100000 以下の素数: 9592 個（割合 9.59%）


## 最大公約数とユークリッドの互除法

紀元前300年頃の「原論」に載っている、世界最古のアルゴリズムの一つです。

**原理**：$a$ を $b$ で割った余りを $r$ とすると、$\gcd(a, b) = \gcd(b, r)$

In [16]:
# Juliaには gcd が標準装備
gcd(1071, 1029)

21

In [17]:
# ユークリッドの互除法を自分で書いてみよう
function my_gcd(a, b)
    while b != 0
        println("gcd($a, $b)")   # 途中経過を表示
        a, b = b, a % b
    end
    return a
end

my_gcd(1071, 1029)

gcd(1071, 1029)
gcd(1029, 42)
gcd(42, 21)


21

In [18]:
# 最小公倍数も標準装備
lcm(12, 18)

36

In [19]:
# 確認：gcd × lcm = 2数の積
gcd(12,18) * lcm(12,18) == 12 * 18

true

## 完全数を探そう

**完全数**とは「自分以外の約数の和が自分自身と等しい数」。

例：$6 = 1 + 2 + 3$

古代ギリシャ時代から知られる神秘的な数です。次の完全数を**探して**みましょう。

In [20]:
# 自分以外の約数の和
proper_sum(n) = sum(divisors(n)) - n

proper_sum(6)

6

In [21]:
# 10000以下の完全数を全部探す!
for n in 2:10000
    if proper_sum(n) == n
        println("$n は完全数! 約数: $(divisors(n))")
    end
end

6 は完全数! 約数: [1, 2, 3, 6]
28 は完全数! 約数: [1, 2, 4, 7, 14, 28]
496 は完全数! 約数: [1, 2, 4, 8, 16, 31, 62, 124, 248, 496]
8128 は完全数! 約数: [1, 2, 4, 8, 16, 32, 64, 127, 254, 508, 1016, 2032, 4064, 8128]


発見した完全数 6, 28, 496, 8128 … 実はすべて $2^{p-1}(2^p - 1)$ の形（$2^p-1$ が素数のとき）をしています。

奇数の完全数が存在するかどうかは、**現在も未解決**です!

## 双子素数を探そう

**双子素数**とは、差が2の素数のペア。例：(3,5), (5,7), (11,13) …

「双子素数は無限にあるか?」も**未解決問題**です。

In [22]:
# 100以下の双子素数を全部探す
for n in 2:100
    if isprime(n) && isprime(n+2)
        println("($n, $(n+2))")
    end
end

(3, 5)
(5, 7)
(11, 13)
(17, 19)
(29, 31)
(41, 43)
(59, 61)
(71, 73)


## フェルマー数

フェルマーは $F_n = 2^{2^n} + 1$ がすべて素数だと予想しました。

$F_0=3,\ F_1=5,\ F_2=17,\ F_3=257,\ F_4=65537$ … 確かに素数。では $F_5$ は?

In [23]:
# フェルマー数を計算して素数判定
for n in 0:5
    F = big(2)^(2^n) + 1   # bigで大きな整数を扱う
    println("F_$n = $F → 素数? $(isprime(F))")
end

F_0 = 3 → 素数? true
F_1 = 5 → 素数? true
F_2 = 17 → 素数? true
F_3 = 257 → 素数? true
F_4 = 65537 → 素数? true
F_5 = 4294967297 → 素数? false


In [24]:
# F_5 を素因数分解してみよう（フェルマーの予想は崩れた!）
factor(big(2)^(2^5) + 1)

641 * 6700417

オイラーが1732年に $F_5 = 641 × 6700417$ を発見し、フェルマーの予想は崩れました。

私たちは**1秒で**その計算を再現できました。これがプログラミングの力です。

## ピタゴラス数を探そう

$a^2 + b^2 = c^2$ を満たす自然数の組 $(a, b, c)$ を**ピタゴラス数**といいます。例：$(3, 4, 5)$

In [25]:
# c ≤ 50 のピタゴラス数を全部探す（a < b とする）
for c in 1:50, b in 1:c, a in 1:b
    if a^2 + b^2 == c^2
        println("($a, $b, $c)")
    end
end

(3, 4, 5)
(6, 8, 10)
(5, 12, 13)
(9, 12, 15)
(8, 15, 17)
(12, 16, 20)
(15, 20, 25)
(7, 24, 25)
(10, 24, 26)
(20, 21, 29)
(18, 24, 30)
(16, 30, 34)
(21, 28, 35)
(12, 35, 37)
(15, 36, 39)
(24, 32, 40)
(9, 40, 41)
(27, 36, 45)
(30, 40, 50)
(14, 48, 50)


## コラッツ予想：世界一有名な未解決問題

どんな自然数 $n$ から始めても、

- 偶数なら **2で割る**
- 奇数なら **3倍して1を足す**

を繰り返すと、**必ず1にたどり着く**（と予想されている）。

1937年に提出されてから、**誰も証明できていません**。しかしJuliaなら数行で実験できます!

In [26]:
# コラッツの操作を繰り返して、1になるまでの道のりを表示
function collatz(n)
    steps = 0
    while n != 1
        n = iseven(n) ? n ÷ 2 : 3n + 1
        print(n, " ")
        steps += 1
    end
    println()
    return steps
end

collatz(6)

3 10 5 16 8 4 2 1 


8

In [27]:
# 27から始めるとかなり長い旅に!
collatz(27)

82 41 124 62 31 94 47 142 71 214 107 322 161 484 242 121 364 182 91 274 137 412 206 103 310 155 466 233 700 350 175 526 263 790 395 1186 593 1780 890 445 1336 668 334 167 502 251 754 377 1132 566 283 850 425 1276 638 319 958 479 1438 719 2158 1079 3238 1619 4858 2429 7288 3644 1822 911 2734 1367 4102 2051 6154 3077 9232 4616 2308 1154 577 1732 866 433 1300 650 325 976 488 244 122 61 184 92 46 23 70 35 106 53 160 80 40 20 10 5 16 8 4 2 1 


111

In [28]:
# 1になるまでのステップ数だけ返すバージョン
function collatz_steps(n)
    steps = 0
    while n != 1
        n = iseven(n) ? n ÷ 2 : 3n + 1
        steps += 1
    end
    return steps
end

# 1〜30のステップ数を実験
for n in 1:30
    println("$n → $(collatz_steps(n)) ステップ")
end

1 → 0 ステップ
2 → 1 ステップ
3 → 7 ステップ
4 → 2 ステップ
5 → 5 ステップ
6 → 8 ステップ
7 → 16 ステップ
8 → 3 ステップ
9 → 19 ステップ
10 → 6 ステップ
11 → 14 ステップ
12 → 9 ステップ
13 → 9 ステップ
14 → 17 ステップ
15 → 17 ステップ
16 → 4 ステップ
17 → 12 ステップ
18 → 20 ステップ
19 → 20 ステップ
20 → 7 ステップ
21 → 7 ステップ
22 → 15 ステップ
23 → 15 ステップ
24 → 10 ステップ
25 → 23 ステップ
26 → 10 ステップ
27 → 111 ステップ
28 → 18 ステップ
29 → 18 ステップ
30 → 18 ステップ


In [29]:
# 1〜10000で一番長くかかる数はどれ?
n_max = argmax(collatz_steps, 1:10000)
println("$n_max が最長で $(collatz_steps(n_max)) ステップ!")

6171 が最長で 261 ステップ!


現在、$2^{68}$（約30京）までの数はすべて1に到達することが**コンピュータで**確認されています。

あなたも今、未解決問題の「検証」に参加したのです!

## Day 2 の演習問題

- まずJuliaのコードで実験してみよう
- ノートブックを保存して、Google Classroomから提出しよう

### 問題1: 約数と完全数

496が完全数であることを、約数をすべて表示して確かめよう。

### 問題2: 素数の実験

1. 2桁の素数は何個あるか（`primes` を使ってもOK）
2. 200以下の双子素数をすべて表示しよう

### 問題3: コラッツの実験

1. 自分の好きな数（例：誕生日の日付）でコラッツの旅を表示してみよう
2. 1〜100の中で一番ステップ数が長い数を見つけよう

### 問題4: チャレンジ問題（カプレカ数）

4桁の数（例：2026）に対して「桁を大きい順に並べた数 − 小さい順に並べた数」を繰り返すと、
必ず **6174** にたどり着きます（カプレカ数）。これを実験で確かめてみよう。

ヒント：`digits(n)` で各桁のリスト、`sort` で並べ替えができます。

## 解答欄

以下のセルに解答を記入してください。

### 問題1の解答

In [30]:
# 496の約数と完全数チェック



### 問題2の解答

In [31]:
# 1. 2桁の素数の個数



In [32]:
# 2. 200以下の双子素数



### 問題3の解答

In [33]:
# 1. 好きな数でコラッツの旅



In [34]:
# 2. 1〜100で最長の数



### 問題4の解答

In [35]:
# カプレカ数の実験



## Day 2 まとめ

今日学習した内容

- **約数・素数**：for文とif文で「全部探す」
- **エラトステネスの篩**：2300年前のアルゴリズムを実装
- **ユークリッドの互除法**：世界最古のアルゴリズム
- **完全数・双子素数・フェルマー数**：数の神秘を自分の手で発見
- **コラッツ予想**：未解決問題も数行で実験できる

数学は「証明」だけでなく「実験」からも始まります。ガウスも、オイラーも、膨大な計算実験から法則を発見しました。

明日は**場合の数**。樹形図を「全部作って」しまいましょう!